base/linalg/triangular.jl

```diff
@@ -149,65 +149,77 @@ istril(A::UnitLowerTriangular) = true
 istriu(A::UpperTriangular) = true
 istriu(A::UnitUpperTriangular) = true
 
-function tril(A::UpperTriangular,k::Integer=0)
+function tril!(A::UpperTriangular,k::Integer=0)
     n = size(A,1)
     if abs(k) > n
         throw(ArgumentError("requested diagonal, $k, out of bounds in matrix of size ($n,$n)"))
     elseif k < 0
-        return UpperTriangular(zeros(A.data))
+        fill!(A.data,0)
+        return A
     elseif k == 0
         return UpperTriangular(diagm(diag(A)))
     else
-        return UpperTriangular(triu(tril(A.data,k)))
+        return UpperTriangular(triu!(tril!(A.data,k)))
     end
 end
+triu!(A::UpperTriangular,k::Integer=0) = UpperTriangular(triu!(triu!(A.data),k))
 
-triu(A::UpperTriangular,k::Integer=0) = UpperTriangular(triu(triu(A.data),k))
-
-function tril(A::UnitUpperTriangular,k::Integer=0)
+function tril!(A::UnitUpperTriangular,k::Integer=0)
     n = size(A,1)
     if abs(k) > n
         throw(ArgumentError("requested diagonal, $k, out of bounds in matrix of size ($n,$n)"))
     elseif k < 0
-        return UnitUpperTriangular(zeros(A.data))
+        fill!(A.data,0)
+        return UpperTriangular(A.data)
     elseif k == 0
         return UnitUpperTriangular(eye(A))
     else
-        return UnitUpperTriangular(triu(tril(A.data,k)))
+        return UnitUpperTriangular(triu!(tril!(A.data,k)))
```

Comment:
this one hurts my head a little - maybe the easiest way to get this right, and get the rest of `tril!(A::UnitUpperTriangular, k)` for almost free would be first set the diagonals of `A.data` to one, then return `tril!(UpperTriangular(A.data), k)` ?